engine/src/distribution/MessageManager.cpp

```diff
@@ -0,0 +1,48 @@
+#include "MessageManager.h"
+#include "utilities/CommonOperations.h"
+
+namespace ral {
+namespace distribution {
+
+int MessageManager::get_total_partition_counts(std::map<std::string, int>& node_count) {
+    int total_count = node_count[node_id];
+    for (auto message : messages_to_wait_for){
+        auto meta_message = input_message_cache->pullCacheData(message);
+        total_count += std::stoi(static_cast<ral::cache::GPUCacheDataMetaData *>(meta_message.get())->getMetadata().get_values()[ral::cache::PARTITION_COUNT]);
+    }
+    return total_count;
+}
+
+void MessageManager::send_total_partition_counts(ral::cache::CacheMachine* graph_output,
+        std::string message_prefix,
+        std::string metadata_label,
+        std::string cache_id,
+        std::map<std::string, int>& node_count) {
```

Comment:
dont pass this is by reference this is bad practice. Return the values you need from the function

Reply:
`node_count` should be a member of the class it's part of the state of the class. It should also be thread safe

Reply:
your comment seems to be addressing some other point. I am saying you cannot pass this map in here by reference.

Reply:
`node_count` should not be passed in at all

Reply:
Fixed.